### PR TITLE
test: twister: parameter in alpha-beta order

### PR DESCRIPTION
**Pull request by hakehuang**
reorde twister according to #38140

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

### CODE REVIEW

File: scripts/twister [hunk 1]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
-        "--verbose",
-        action="count",
-        default=0,
-        help="Emit debugging information, call multiple times to increase "
-             "verbosity")
+        "-H", "--footprint-threshold", type=float, default=5,
+        help="When checking test case footprint sizes, warn the user if "
+             "the new app size is greater then the specified percentage "
+             "from the last release. Default is 5. 0 to warn on any "
+             "increase on app size")
```
Comment (PerMac) [inline scripts/twister:563]:
```suggestion
             "increase on app size.")
```

File: scripts/twister [hunk 2]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
+
     parser.add_argument(
         "-i", "--inline-logs", action="store_true",
         help="Upon test failure, print relevant log data to stdout "
              "instead of just a path to it")
```
Comment (PerMac) [inline scripts/twister:568]:
```suggestion
             "instead of just a path to it.")
```

File: scripts/twister [hunk 3]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
-
-    case_select.add_argument(
-        "-F",
-        "--load-tests",
-        metavar="FILENAME",
-        action="store",
-        help="Load list of tests and platforms to be run from file.")
+        "-j", "--jobs", type=int,
+        help="Number of jobs for building, defaults to number of CPU threads, "
+             "overcommited by factor 2 when --build-only")
```
Comment (PerMac) [inline scripts/twister:573]:
```suggestion
             "overcommited by factor 2 when --build-only.")
```

File: scripts/twister [hunk 4]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
-        "-b", "--build-only", action="store_true",
-        help="Only build the code, do not execute any of it in QEMU")
-
-    test_or_build.add_argument(
-        "--test-only", action="store_true",
-        help="""Only run device tests with current artifacts, do not build
-             the code""")
+        "-K", "--force-platform", action="store_true",
+        help="""Force testing on selected platforms,
+        even if they are excluded in the test configuration (testcase.yaml)"""
```
Comment (PerMac) [inline scripts/twister:583]:
```suggestion
        even if they are excluded in the test configuration (testcase.yaml)."""
```

File: scripts/twister [hunk 5]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
 
-    parser.add_argument(
-        "--filter", choices=['buildable', 'runnable'],
-        default='buildable',
-        help="""Filter tests to be built and executed. By default everything is
-        built and if a test is runnable (emulation or a connected device), it
-        is run. This option allows for example to only build tests that can
-        actually be run. Runnable is a subset of buildable.""")
+    parser.add_argument("--list-tags", action="store_true",
+                        help="list all tags in selected tests")
```
Comment (PerMac) [inline scripts/twister:592]:
```suggestion
                        help="List all tags occuring in selected tests.")
```

File: scripts/twister [hunk 6]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
 
+    parser.add_argument("--log-file", metavar="FILENAME", action="store",
+                        help="log also to file")
```
Comment (PerMac) [inline scripts/twister:595]: This msg is too cryptic.  "Specify a file where to save logs." ?

File: scripts/twister [hunk 7]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
-        help="Show footprint statistics and deltas since last release."
-    )
-    parser.add_argument(
-        "-H", "--footprint-threshold", type=float, default=5,
-        help="When checking test case footprint sizes, warn the user if "
-             "the new app size is greater then the specified percentage "
-             "from the last release. Default is 5. 0 to warn on any "
-             "increase on app size")
+        "-N", "--ninja", action="store_true",
+        help="Use the Ninja generator with CMake")
```
Comment (PerMac) [inline scripts/twister:608]:
```suggestion
        help="Use the Ninja generator with CMake.")
```

File: scripts/twister [hunk 8]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
-        "-T", "--testcase-root", action="append", default=[],
-        help="Base directory to recursively search for test cases. All "
-             "testcase.yaml files under here will be processed. May be "
-             "called multiple times. Defaults to the 'samples/' and "
-             "'tests/' directories at the base of the Zephyr tree.")
 
-    board_root_list = ["%s/boards" % ZEPHYR_BASE,
-                       "%s/scripts/pylib/twister/boards" % ZEPHYR_BASE]
+    parser.add_argument("--overflow-as-errors", action="store_true",
+                        help="Treat RAM/SRAM overflows as errors")
```
Comment (PerMac) [inline scripts/twister:627]:
```suggestion
                        help="Treat RAM/SRAM overflows as errors.")
```

File: scripts/twister [hunk 9]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
     parser.add_argument(
-        "-S", "--enable-slow", action="store_true",
-        help="Execute time-consuming test cases that have been marked "
-             "as 'slow' in testcase.yaml. Normally these are only built.")
+        "--quarantine-list",
+        metavar="FILENAME",
+        help="Load list of test scenarios under quarantine. The entries in "
+             "the file need to correspond to the test scenarios names as in"
+             "corresponding tests .yaml files. These scenarios"
+             "will be skipped with quarantine as the reason")
```
Comment (PerMac) [inline scripts/twister:668]:
```suggestion
        help="Load list of test scenarios under quarantine. The entries in "
             "the file need to correspond to the test scenarios names as in "
             "corresponding tests .yaml files. These scenarios "
             "will be skipped with quarantine as the reason.")
```

File: scripts/twister [hunk 10]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
 
     parser.add_argument(
-        "-K", "--force-platform", action="store_true",
-        help="""Force testing on selected platforms,
-        even if they are excluded in the test configuration (testcase.yaml)"""
-    )
+        "--quarantine-verify",
+        action="store_true",
+        help="Use the list of test scenarios under quarantine and run them"
+             "to verify their current status")
```
Comment (PerMac) [inline scripts/twister:674]:
```suggestion
        help="Use the list of test scenarios under quarantine and run them"
             "to verify their current status.")
```

File: scripts/twister [hunk 11]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
 
     parser.add_argument(
-        "--disable-unrecognized-section-test", action="store_true",
-        default=False,
-        help="Skip the 'unrecognized section' test.")
+        "-r", "--release", action="store_true",
+        help="Update the benchmark database with the results of this test "
+             "run. Intended to be run by CI when tagging an official "
+             "release. This database is used as a basis for comparison "
+             "when looking for deltas in metrics such as footprint")
```
Comment (PerMac) [inline scripts/twister:681]:
```suggestion
             "when looking for deltas in metrics such as footprint.")
```

File: scripts/twister [hunk 12]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
+    )
 
-                        E.g "twister --device-testing
-                        --device-serial-pty <script>
-                        """)
+    parser.add_argument(
+        "-t", "--tag", action="append",
+        help="Specify tags to restrict which tests to run by tag value. "
+             "Default is to not do any tag filtering. Multiple invocations "
+             "are treated as a logical 'or' relationship")
```
Comment (PerMac) [inline scripts/twister:726]:
```suggestion
             "are treated as a logical 'or' relationship.")
```

File: scripts/twister [hunk 13]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
 
-    parser.add_argument("--generate-hardware-map",
-                        help="""Probe serial devices connected to this platform
-                        and create a hardware map file to be used with
-                        --device-testing
-                        """)
+    parser.add_argument("--timestamps",
+                        action="store_true",
+                        help="Print all messages with time stamps")
```
Comment (PerMac) [inline scripts/twister:730]:
```suggestion
                        help="Print all messages with time stamps.")
```

File: scripts/twister [hunk 14]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
-    parser.add_argument("--persistent-hardware-map", action='store_true',
-                        help="""With --generate-hardware-map, tries to use
-                        persistent names for serial devices on platforms
-                        that support this feature (currently only Linux).
-                        """)
+    parser.add_argument(
+        "-u",
+        "--no-update",
+        action="store_true",
+        help="do not update the results of the last run of the twister run")
```
Comment (PerMac) [inline scripts/twister:736]:
```suggestion
        help="Do not update the results of the last run of twister.")
```

File: scripts/twister [hunk 15]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
-                        help="""Load hardware map from a file. This will be used
-                        for testing on hardware that is listed in the file.
-                        """)
+    parser.add_argument(
+        "-v",
+        "--verbose",
+        action="count",
+        default=0,
+        help="Emit debugging information, call multiple times to increase "
+             "verbosity")
```
Comment (PerMac) [inline scripts/twister:744]:
```suggestion
             "verbosity.")
```

File: scripts/twister [hunk 16]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
 
-    parser.add_argument("--pre-script",
-                        help="""specify a pre script. This will be executed
-                        before device handler open serial port and invoke runner.
-                        """)
+    parser.add_argument("-W", "--disable-warnings-as-errors", action="store_true",
+                        help="Treat warning conditions as errors")
```
Comment (PerMac) [inline scripts/twister:747]:
```suggestion
                        help="Treat warning conditions as errors.")
```

File: scripts/twister [hunk 17]
```diff
@@ -683,61 +771,38 @@ structure in the main Zephyr tree: boards/<arch>/<board_name>/""")
...
 
     parser.add_argument(
-        "--enable-lsan", action="store_true",
-        help="""Enable leak sanitizer to check for heap memory leaks.
-        Libasan needs to be installed on the host. This option only
-        works with host binaries such as those generated for the native_posix
-        configuration and when --enable-asan is given.
-        """)
+        "-X", "--fixture", action="append", default=[],
+        help="Specify a fixture that a board might support")
```
Comment (PerMac) [inline scripts/twister:791]:
```suggestion
        help="Specify a fixture that a board might support.")
```

**Review by PerMac — DISMISSED** [review]

File: scripts/twister [hunk 18]
```diff
@@ -322,48 +343,112 @@ Artificially long but functional example:
...
+                        --device-testing.
+                        E.g --generate-hardware-map map.yml
+                        """)
+
+    run_group_option.add_argument(
+        "--device-testing", action=SerialAction,
+        help="Test on device directly. Specify the serial device to "
+             "use with the --device-serial option."
+             "E.g. --device-testing --device-serial /dev/ttyACM0"
+             "or --device-testing --hardware-map map.yml")
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/twister:386]: ok, Thanks

File: scripts/twister [hunk 19]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
+                        help="Tool to use to generate coverage report.")
 
-    case_select.add_argument("--list-test-duplicates", action="store_true",
-                             help="""List tests with duplicate identifiers.
+    parser.add_argument(
+        "-D", "--all-deltas", action="store_true",
+        help="Show all footprint deltas, positive or negative. Implies "
+             "--footprint-threshold=0. By default it compares with"
+             "scripts/release/twister_last_release.csv"
+             "or use --compare-report or --last-metrics")
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/twister:489]: ok, thanks

File: scripts/twister [hunk 20]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
+        built and if a test is runnable (emulation or a connected device), it
+        is run. This option allows for example to only build tests that can
+        actually be run. Runnable is a subset of buildable.""")
 
-    parser.add_argument("-W", "--disable-warnings-as-errors", action="store_true",
-                        help="Treat warning conditions as errors")
+    parser.add_argument("--force-toolchain", action="store_true",
+                        help="Do not filter based on toolchain, use the set "
+                             " toolchain unconditionally."
+                             " toolchain defined in board/platform yml file")
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/twister:551]: ok

File: scripts/twister [hunk 21]
```diff
@@ -373,290 +458,293 @@ Artificially long but functional example:
...
 
-    parser.add_argument("--overflow-as-errors", action="store_true",
-                        help="Treat RAM/SRAM overflows as errors")
+    parser.add_argument("--gcov-tool", default=None,
+                        help="Path to the gcov tool to use for code coverage "
+                             "reports. default path ${ZEPHYR_SDK_INSTALL_DIR}/"
+                             "x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcov")
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/twister:556]: if no parameter is given the path will be ${ZEPHYR_SDK_INSTALL_DIR}/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcov, Let me rephase it

File: scripts/twister (hunk 1 above)
Comment (hakehuang) [reply scripts/twister:563]: ok

File: scripts/twister (hunk 3 above)
Comment (hakehuang) [reply scripts/twister:573]: ok

File: scripts/twister (hunk 4 above)
Comment (hakehuang) [reply scripts/twister:583]: ok

File: scripts/twister (hunk 5 above)
Comment (hakehuang) [reply scripts/twister:592]: ok

File: scripts/twister (hunk 6 above)
Comment (hakehuang) [reply scripts/twister:595]: ok

File: scripts/twister (hunk 7 above)
Comment (hakehuang) [reply scripts/twister:608]: ok

File: scripts/twister (hunk 8 above)
Comment (hakehuang) [reply scripts/twister:627]: ok

File: scripts/twister (hunk 9 above)
Comment (hakehuang) [reply scripts/twister:668]: ok

File: scripts/twister (hunk 10 above)
Comment (hakehuang) [reply scripts/twister:674]: ok

File: scripts/twister (hunk 12 above)
Comment (hakehuang) [reply scripts/twister:726]: ok

File: scripts/twister (hunk 13 above)
Comment (hakehuang) [reply scripts/twister:730]: ok

File: scripts/twister (hunk 14 above)
Comment (hakehuang) [reply scripts/twister:736]: ok

File: scripts/twister (hunk 15 above)
Comment (hakehuang) [reply scripts/twister:744]: ok

File: scripts/twister (hunk 16 above)
Comment (hakehuang) [reply scripts/twister:747]: ok

File: scripts/twister (hunk 17 above)
Comment (hakehuang) [reply scripts/twister:791]: ok

File: scripts/twister [hunk 22]
```diff
@@ -245,69 +243,92 @@ def export_tests(filename, tests):
...
+            const=const,
+            default=default,
+            required=required,
+            help=help)
+    def __call__(self, parser, namespace, values, option_string=None):
+        if not namespace.device_serial  and \
+           not namespace.device_serial_pty and \
+           not namespace.hardware_map:
+            print(namespace)
+            parser.error("Missing opts for --device-testing"
```
[the remark this answers is not in the document]
Comment (PerMac) [reply scripts/twister]: I see. Then I think in help for  "--outdir" there are '' used inconsistently with other section

[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/twister]: right, It is reused from former code, let me correct it

File: scripts/twister [hunk 23]
```diff
@@ -245,69 +243,92 @@ def export_tests(filename, tests):
                 logger.error("{} can't be exported: ".format(test))
 
 
+class SerialAction(argparse.Action):
+    def __init__(self,
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/twister:247]: @PerMac , for this I write a mail to you, I can't find a good solution for this
```
I try to implement the dependency on –device-test on –device-serial / --device-serial-pty / --hardware-map

https://github.com/hakehuang/zephyr/blob/e28c7bc8d96f96836ab05ae2f8640af98624dd2e/scripts/twister#L248

but I meet a problem, above code only works when you command line as below

./scripts/twister –device-serial /dev/ttyACM0 –device-test

But not work when you do below, which I believe is most of our default scripts does

./scripts/twister –device-test –device-serial /dev/ttyACM0

```

**Review by nashif — CHANGES_REQUESTED** [review]
this change is very difficult to review, the title is about alphabetical order but it has lots of other unrelated changes all compressed into one commit.
- IMO instead of focusing on alphabetical order, focus on grouping, I doubt alphabetical ordering would improve the help if related options are all over the place.
- Address language and syntax of option help separately from functional and logical changes.

**Comment by hakehuang** [comment]
@anasnashif actually I do grouping in this PR as well. and I move the grouping options at the begining of twister options, and keep them in order as possible. 
As to alph-beta orderring, I think it is helpful when we adding new opts to code, current the opts are added at will. IMO, an orderred system can be more reliable.
also there are some changes to address the language and syntax issues, and some helper enhancements.

Do you mean I need separated above three changes to different commit?

**Review by nashif — APPROVED** [review]

**Review by PerMac — APPROVED** [review]

**Pull request merged**